### PR TITLE
Implement group/team info caching per issue #367

### PR DESCRIPTION
**Pull request by cdoremus**
This PR implements an application cache key in the Redux store to hold group/team data that is persisted using `redux-persist`. It is an implementation of issue #367.

### CODE REVIEW

**Comment by cdoremus** [comment]
@nickoneill what timeout value would you suggest for the cache?

**Comment by nickoneill** [comment]
not sure yet, let's start at 24h

**Comment by nickoneill** [comment]
How's this going @cdoremus? Anything I can do to help?

**Comment by cdoremus** [comment]
@nickoneill Sorry, but I've been distracted by job search and holiday responsibilities. Still, I have done some work that I should be able to push this evening or tomorrow.

**Comment by nickoneill** [comment]
No worries! Just checking in. Let me know if I can help or review.

**Comment by cdoremus** [comment]
The commit I pushed today has failing tests because of some `console.log()` calls I added for debugging purposes. These will be removed when the code is ready for review. If you run the tests using `yarn test:nolint`, all tests pass.

**Comment by cdoremus** [comment]
Making progress... 

I still need to change `currentGroupId: string` prop to `currentGroup: Group` to allow display of group name in the `SidebarHeader` and manually test some corner cases.

The unit tests now pass.

**Comment by nickoneill** [comment]
🎇

**Comment by nickoneill** [comment]
Awesome! I am just starting to go through this and my first pass worked fine, but deleting the appCache object gets me stuck into this state where my group isn't cached and it won't refetch.

I'm trying to figure out where the cache timestamp is stored but I can't figure it out - how have you been resetting this to try it fresh?

**Comment by cdoremus** [comment]
@nickoneill How are you deleting the `appCache` object? I want to be able to reproduce the bug you found.

All group data gets stored in the `reduxPersist:appCache` key in local storage. When I delete that key, the app continues to function OK. The `appCache.groups` object contains an array of `CacheableGroup` objects. Each `CacheableGroup` contains properties `group: Group` and `timestamp: number` (milliseconds). 

When a group is requested, the `appCache.groups` array is searched by group id and if a group is found  the timestamp of the `CacheableGroup` that it is in iis checked against the current timestamp to see if the cached object has timed out. (see `redux/cache/asyncActionCreator.cacheGroup()`). If the group has timed out, an API call is made to update the group object of the `CacheableGroup` and the timestamp is set to the current timestamp. 

If a group is not found a new Group is created from an API call and added to a new `CacheableGroup` with a current timestamp. This new `CacheableGroup` is then added to the `appCache.groups` Redux state object.

I looked at the code again and have started refactoring the `GroupPageContainer` and `GroupPage` to clarify and simplify things. I also need to make sure the cache timeout is checked in `GroupPage`. I hope to check this stuff in as soon as time allows.

**Comment by nickoneill** [comment]
I just deleted the `reduxPersist:appCache` key in localstorage through chrome. It populates again when I load another group page but it's empty (`{"groups":[]}`), it never seems to pick up that particular group information.

**Comment by cdoremus** [comment]
@nickoneill I just discovered I need to change all the `currentGroup` references throughout from type `Group` to `CachableGroup`. Expect another commit for this, but please review what I committed today. TIA

**Comment by cdoremus** [comment]
I've done the refactoring and everything should be ready to go (I think :smiley_cat: ).

I have to say that TypeScript really makes these kind of refactorings easy!

File: src/components/call/FetchCall.tsx [hunk 1]
```diff
@@ -250,15 +250,15 @@ export default class FetchCall extends React.Component<Props, State> {
...
-      return <h3 className="call__outcomes__header">Getting your next contact...</h3>;      
+      return <h3 className="call__outcomes__header">Getting your next contact...</h3>;
     } else {
       return (
         <blockquote>
           <h2 className="call__outcomes__header">All done for today!</h2>
           {/*tslint:disable-next-line:max-line-length*/}
           <p>Looks like we're all out of calls to make for today, or we're outside of normal calling hours (9am-9pm in the local time zone). Come back tomorrow for more calls!</p>
         </blockquote>
-      );
```
Comment (nickoneill) [inline src/components/call/FetchCall.tsx:126]: Sorry about these, I don't know why VS code doesn't trim them like any sane software

File: src/components/groups/GroupPage.tsx [hunk 2]
```diff
@@ -26,63 +24,84 @@ interface Props extends RouteProps {
...
-    if (this.state.pageGroup) {
-      if (newProps.match.params.groupid !== this.state.pageGroup.id) {
-        this.setState({ loadingState: GroupLoadingState.LOADING });
-        this.getGroupDetails(newProps.match.params.groupid);
-      }      
+  setGroup(props: Props) {
+    let cgroup = this.state.pageGroupState;
+    // Call the API when page first loads and group
+    // is not in the cache or if the cached value has timed out
+    if (!cgroup || hasGroupCacheTimeoutExceeded({timestamp: cgroup.timestamp})) {
```
Comment (nickoneill) [inline src/components/groups/GroupPage.tsx:62]: so the page is responsible for expiring a group? Feels like this might be duplicated on multiple pages in the future (whenever we use a group) but probably OK for the moment.

Not sure where I can imagine this going. Since the getter for groups is really just the state object and not some function that you can insert timeout checking.

**Review by nickoneill — CHANGES_REQUESTED** [review]
Overall great stuff! One specific comment inline without any particular changes requested, but curious on your opinion.

One thing I noticed: if you have no cached groups and you hit a call detail page (`team/nick/2`), the group isn't fetched and cached. The sidebar still lists the groups issues but the links are all broken since there's no actual group id to construct a proper group url.

Ideas here? This might touch on that other code comment I had as well now that requesting the group cache is on more than one page.

**Comment by nickoneill** [comment]
Gonna resolve the conflict here and pull this in as a start so I can base some future work on this.

**Comment by nickoneill** [comment]
Why I had to merge the same conflict twice is beyond me.
